association.go

```diff
@@ -85,11 +94,16 @@ func (association *Association) Replace(values ...interface{}) error {
 					association.Error = rel.Field.Set(association.DB.Statement.Context, reflectValue, reflect.Zero(rel.Field.FieldType).Interface())
 				}
 
-				for _, ref := range rel.References {
-					updateMap[ref.ForeignKey.DBName] = nil
+				if association.Unscope {
+					modelValue := reflect.New(rel.FieldSchema.ModelType).Interface()
```

Comment:
no, I will add it

Reply:
Maybe it shouldn't work on `BelongsTo`, I removed the code.
Now it works only on `HasOne/HasMany`

Reply:
I rarely use Association, I'm not sure if I'm right.

> A belongs to association sets up a one-to-one connection with another model, such that each instance of the declaring model “belongs to” one instance of the other model.

https://gorm.io/docs/belongs_to.html#Belongs-To
To behave similarly to HasMany, in BelongsTo it should do the following:
Delete:
1. set User.CompanyID to Null 
2. delete Company
Replace:
1. set User.CompanyID to new CompanyID
2. delete Company

Reply:
Actually it (Replace) is:
1. delete old Company
2. insert new Company and update User.CompanyID

It is not the same as `has one`...
I won't admit I want to be lazy 🙄

Reply:
> I rarely use Association, I'm not sure if I'm right.
> 
> > A belongs to association sets up a one-to-one connection with another model, such that each instance of the declaring model “belongs to” one instance of the other model.
> 
> https://gorm.io/docs/belongs_to.html#Belongs-To To behave similarly to HasMany, in BelongsTo it should do the following: Delete:
> 
> 1. set User.CompanyID to Null
> 2. delete Company
>    Replace:
> 3. set User.CompanyID to new CompanyID
> 4. delete Company
> 
> It might just not work with Many2Many

I will try to implement it

Reply:
This is how I understand Unscoped's behavior as a user.
You don't have to implement it unless you think so too. In fact, if our understanding is different, we can open an issue for users to discuss it, so as to avoid making useless features.

Reply:
> This is how I understand Unscoped's behavior as a user. You don't have to implement it unless you think so too. In fact, if our understanding is different, we can open an issue for users to discuss it, so as to avoid making useless features.

We're on the same mind, but I have no idea about where to implement it. `Replace` or `saveAssociation` ?

Reply:
> > This is how I understand Unscoped's behavior as a user. You don't have to implement it unless you think so too. In fact, if our understanding is different, we can open an issue for users to discuss it, so as to avoid making useless features.
> 
> We're on the same mind, but I have no idea about where to implement it. `Replace` or `saveAssociation` ?

This is indeed controversial, according to the current implementation, I think it should be implemented in `Replace`. Although `saveAssociation` seems appropriate, but since the logic of `Clear` is currently placed on `Replace`, it will become more troublesome if you want to implementation in `saveAssociation`

https://github.com/go-gorm/gorm/blob/master/association.go#L77

Reply:
I'm done, please review it. @a631807682